### ISSUE TITLE
Efficiency of certain operations

### ISSUE THREAD

**Issue by Cheappie** [RoaringBitmap/RoaringBitmap#398]
Hi,

recently while calculating AND cardinality I have hit the wall, unexpectedly performance did drop significantly. After some research and debugging I came up with code which reproduce such situation. Would it be correct assumption that whenever run container appear in bitmap, it will significantly slowdown operations ?

`
    long rangeEnd = 100_000_000;

    RoaringBitmap r1 = new RoaringBitmap();
    r1.add(0L, rangeEnd);

    Random rand = new Random();
    RoaringBitmap temp = new RoaringBitmap();
    for (int i = 0; i < rangeEnd; i++) {
        if (rand.nextBoolean()) {
            temp.add(i);
        }
    }
    RoaringBitmap r2 = RoaringBitmap.bitmapOf(temp.toArray());

    System.out.println("r1 bytes: " + r1.getSizeInBytes());
    System.out.println("r2 bytes: " + r2.getSizeInBytes());

    long cards = 0;
    List<Long> timer = new ArrayList<>();
    for (int i = 0; i < 100; i++) {
        long st = System.currentTimeMillis();
        cards += RoaringBitmap.andCardinality(r1, r2); // ~ 60 to 80 times slower than below
        // cards += RoaringBitmap.and(r1,r2).getCardinality(); // faster
        timer.add(System.currentTimeMillis() - st);
    }
    System.out.println("Millis: " + timer.stream().mapToLong(Long::longValue).summaryStatistics());
    System.out.println(cards);
`

**Comment by richardstartin** [comment]
@Cheappie looks sensible. Do you want to make a PR with the change?

**Comment by Cheappie** [comment]
If you don't mind I would be glad if you could create PR to apply that change. I don't know what processes you undertake for introducing changes in repository. It makes me really happy that this function will perform better.

Could you tell me what would me approximate time to let that change reach mvn central, probably new release would be necessary ?

**Comment by Cheappie** [comment]
That's only my guess

**RunContainer.intersects(BitmapContainer x)**
has comment '// TODO: this is probably not optimally fast', because it does brute-force contains check. Where it could reuse bitmapContainer.intersects(int minimum, int supremum) function

**Comment by richardstartin** [comment]
Yes, `RunContainer.intersects(BitmapContainer)` can also be sped up.

**Comment by richardstartin** [comment]
I suspect `RunContainer.and(BitmapContainer)` is sub-optimal and it would be better either to convert to a bitmap container or to iterate over the bitmap container within each run.

https://github.com/RoaringBitmap/RoaringBitmap/blob/master/RoaringBitmap/src/main/java/org/roaringbitmap/RunContainer.java#L353

**Comment by lemire** [comment]
> I suspect RunContainer.and(BitmapContainer) is sub-optimal and it would be better either to convert to a bitmap container or to iterate over the bitmap container within each run.

It is. I think I wrote it and figured that it would be "good enough". It probably can be made 2x faster, at least. It might be best to create a specific issue to track this one thing.
